include/nbl/asset/CCPUMeshPackerV2.h

```diff
@@ -0,0 +1,222 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef __NBL_ASSET_C_CPU_MESH_PACKER_V2_H_INCLUDED__
+#define __NBL_ASSET_C_CPU_MESH_PACKER_V2_H_INCLUDED__
+
+#include <nbl/asset/ICPUMesh.h>
+#include <nbl/asset/IMeshPacker.h>
+
+namespace nbl
+{
+namespace asset
+{
+
+// eventually I will rename it
+template <typename MDIStructType = DrawElementsIndirectCommand_t>
+class CCPUMeshPackerV2 final : public IMeshPacker<ICPUMeshBuffer, MDIStructType>
+{
+    using base_t = IMeshPacker<ICPUMeshBuffer, MDIStructType>;
+    using Triangle = typename base_t::Triangle;
+    using TriangleBatch = typename base_t::TriangleBatch;
+
+public:
+    using AllocationParams = MeshPackerBase::AllocationParamsCommon;
+
+    struct AttribAllocParams
+    {
+        size_t offset = INVALID_ADDRESS;
+        size_t size = 0ull;
+    };
+
+	struct ReservedAllocationMeshBuffers
+    {
+        uint32_t mdiAllocationOffset;
+        uint32_t mdiAllocationReservedSize;
+        uint32_t indexAllocationOffset;
+        uint32_t indexAllocationReservedSize;
+        AttribAllocParams attribAllocParams[SVertexInputParams::MAX_VERTEX_ATTRIB_COUNT];
+
+        inline bool isValid()
+        {
+            return this->mdiAllocationOffset != core::GeneralpurposeAddressAllocator<uint32_t>::invalid_address;
+        }
+    };
+
+    struct PackedMeshBuffer
+    {
+        core::smart_refctd_ptr<video::IGPUBuffer> MDIDataBuffer;
+        core::smart_refctd_ptr<video::IGPUBuffer> vertexBuffer;
+        core::smart_refctd_ptr<video::IGPUBuffer> indexBuffer;
+
+        inline bool isValid()
+        {
+            return this->MDIDataBuffer->getPointer() != nullptr;
+        }
+    };
+
+public:
+    CCPUMeshPackerV2(const AllocationParams& allocParams, uint16_t minTriangleCountPerMDIData = 256u, uint16_t maxTriangleCountPerMDIData = 1024u);
+
+	template <typename Iterator>
+	bool alloc(ReservedAllocationMeshBuffers* rambOut, const Iterator begin, const Iterator end);
```

Comment:
why is this a `bool` wheareas the V1 packer just returns the `ReservedAllocationMeshBuffers`

Reply:
return value indicates if allocation was successfull, whereas `ReservedAllocationMeshBuffers` is saved to `rambOut` array. It is because `alloc` of `CCPUMeshPackerV2` returns multiple `ReservedAllocationMeshBuffers`

Reply:
alternatively this function may return `core::vector<ReservedAllocationMeshBuffers>`

Reply:
ok thanks for explanation... keep as is ;)

but then why does V1 API not look like the new V2 ?

Reply:
> but then why does V1 API not look like the new V2 ?

It doesn't need to I suppose.. unless you really want them to be consistent?

sorry, I missed this question

Reply:
the question is why does V1 return only one reserved allocation and no `bool` to notify of success?

Reply:
V1 returns `ReservedAllocationMeshBuffers` struct, which has the `isValid` function. `isValid` indicates if allocation was successful

Reply:
Ok I'm starting to see the reason for the difference.

Because the individual allocation for a meshbuffer is dictated by `baseVertex` or `firstIndex` of the MDI, its only necessary to know about the MDIs allocated (`ReservedAllocationMeshBuffers` has this info about the MDI range) to "extract" the rest of the parameters.

Whereas here the allocations for the vertex and index data can be very independent per meshbuffer and not follow any pattern (although one can still extract the index allocation from the first MDI struct's `firstIndex`).